rust/bundle-builder/src/main.rs

```diff
@@ -0,0 +1,271 @@
+            .context(RunServerSnafu)
+    });
+
+    future::select(reflector, server).await.factor_first().0
+}
+
+#[derive(Snafu, Debug)]
+#[snafu(module)]
+enum BundleError {
+    #[snafu(display("ConfigMap is missing required metadata"))]
+    ConfigMapMetadataMissing,
+
+    #[snafu(display("file {file_name:?} in {config_map} is too large ({file_size} bytes)"))]
+    FileSizeOverflow {
+        source: TryFromIntError,
+        config_map: ObjectRef<ConfigMap>,
+        file_name: String,
+        file_size: usize,
+    },
+
+    #[snafu(display("failed to add file {file_name:?} from {config_map} to tarball"))]
+    AddFileToTarball {
+        source: std::io::Error,
+        config_map: ObjectRef<ConfigMap>,
+        file_name: String,
+    },
+
+    #[snafu(display("failed to build tarball"))]
+    BuildTarball { source: std::io::Error },
+}
+
+impl BundleError {
+    fn to_http_response(&self) -> impl IntoResponse {
+        (
+            http::StatusCode::INTERNAL_SERVER_ERROR,
+            "failed to build bundle, see opa-bundle-builder logs for more details",
+        )
+    }
+}
+
+async fn build_bundle(store: Store<ConfigMap>) -> Result<Vec<u8>, BundleError> {
+    use bundle_error::*;
+    fn file_header(
+        config_map: &ObjectRef<ConfigMap>,
+        file_name: &str,
+        data: &[u8],
+    ) -> Result<tar::Header, BundleError> {
+        let mut header = tar::Header::new_gnu();
+        header.set_mode(0o644);
+        let file_size = data.len();
+        header.set_size(
+            file_size
+                .try_into()
+                .with_context(|_| FileSizeOverflowSnafu {
+                    config_map: config_map.clone(),
+                    file_name,
+                    file_size,
+                })?,
+        );
+        header.set_entry_type(tar::EntryType::Regular);
+        header.set_cksum();
+        Ok(header)
+    }
+
+    info!("building bundle");
+    let mut tar = tar::Builder::new(GzEncoder::new(Vec::new(), flate2::Compression::default()));
+    for cm in store.state() {
+        let ObjectMeta {
+            name: Some(cm_ns),
+            namespace: Some(cm_name),
+            ..
+        } = &cm.metadata
+        else {
+            return ConfigMapMetadataMissingSnafu.fail();
+        };
+        let cm_ref = ObjectRef::from_obj(&*cm);
+        for (file_name, data) in cm.data.iter().flatten() {
+            let mut header = file_header(&cm_ref, file_name, data.as_bytes())?;
+            tar.append_data(
+                &mut header,
+                format!("configmap/{cm_ns}/{cm_name}/{file_name}"),
+                data.as_bytes(),
+            )
+            .with_context(|_| AddFileToTarballSnafu {
+                config_map: cm_ref.clone(),
+                file_name,
+            })?;
+        }
+    }
+    let tar = tar
+        .into_inner()
+        .context(BuildTarballSnafu)?
+        .finish()
+        .context(BuildTarballSnafu)?;
+    info!("finished building bundle");
```

Comment:
Can we please also include all the `format!("configmap/{cm_ns}/{cm_name}/{file_name}")` that we bundled? So that users can see what rego rules actually were loaded?

Reply:
That's gonna be a loooong list, but sure.

Reply:
17173f93a62842bf9f1d685ef1cd5dde71fa5705